adserver/middleware.py

```diff
@@ -137,8 +99,110 @@ def _get_ip_address(self, request):
                 )
-                client_ip = None
-        if not client_ip:
-            client_ip = request.META.get("REMOTE_ADDR", "")
 
-        return client_ip
+        return ip
+
+
+class CloudflareIpAddressMiddleware(IpAddressMiddleware):
+
+    """
+    Sets request.ip_address from Cloudflare headers.
+
+    See: https://developers.cloudflare.com/fundamentals/get-started/http-request-headers
+
+    SECURITY NOTE: This middleware *SHOULD BE DISABLED* if not running on Cloudflare.
+        Otherwise these headers are spoofable and could result in invalid traffic.
+    """
+
+    PROVIDER_NAME = "Cloudflare"
+    IP_HEADER = "CF-Connecting-IP"
+
+    def get_ip_address(self, request):
+        ip = super().get_ip_address(request)
+        cf_ip = request.headers.get(self.IP_HEADER, None)
+
+        if cf_ip:
+            try:
+                ipaddress.ip_address(cf_ip)
+                ip = cf_ip
+            except ValueError:
+                # This should be OK to log since this is a spoofed address
+                log.warning(
+                    "Invalid IP in Cloudflare headers. "
+                    "This header is probably untrustworthy so the CloudflareIpAddressMiddleware should be disabled. "
+                    "%s=%s",
+                    self.IP_HEADER,
+                    cf_ip,
+                )
+
+        return ip
+
+
+class GeoIpMiddleware:
+
+    """
+    Sets a request.geo dictionary onto the request object.
+
+    This middleware doesn't have any provider and just sets null
+    for all the geolocation data. It isn't useful except in development/testing.
+
+    In production, you probably want to use the CloudflareGeoIpMiddleware
+    or the GeoIpDbMiddleware depending on your setup.
+    """
+
+    PROVIDER_NAME = "None"
+
+    def __init__(self, get_response):
+        """One-time configuration and initialization."""
+        self.get_response = get_response
+
+    def __call__(self, request):
+        """Sets the geolocation data of the user onto the request object."""
+        request.geo = self.get_geoip(request)
+        response = self.get_response(request)
+
+        if settings.DEBUG or request.user.is_staff:
+            response["X-Adserver-GeoIp-Provider"] = self.PROVIDER_NAME
+        return response
+
+    def get_geoip(self, request):
+        return GeolocationData()
+
+
+class CloudflareGeoIpMiddleware(GeoIpMiddleware):
+
+    """
+    Sets request.geo from Cloudflare headers.
+
+    See: https://support.cloudflare.com/hc/en-us/articles/200168236-Configuring-Cloudflare-IP-Geolocation
+
+    SECURITY NOTE: This middleware *SHOULD BE DISABLED* if not running on Cloudflare.
+        Otherwise these headers are spoofable and could result in invalid traffic.
+        Also, Cloudflare IP Geolocation must be enabled.
+    """
+
+    COUNTRY_HEADER = "CF-IPCountry"
+    PROVIDER_NAME = "Cloudflare"
+
+    def get_geoip(self, request):
+        geo = super().get_geoip(request)
```

Comment:
This is basically just getting an empty `GeolocationData` instance? Or are there cases where it might already have some data set?

Reply:
It should always be empty.